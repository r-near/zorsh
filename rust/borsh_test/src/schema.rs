use borsh::schema::{BorshSchemaContainer, Definition, Fields};
use borsh::BorshDeserialize;
use std::{collections::BTreeMap, fs::File, io::Write};

struct SchemaGenerator<'a> {
    container: &'a BorshSchemaContainer,
    generated: BTreeMap<String, String>,
}

impl<'a> SchemaGenerator<'a> {
    fn new(container: &'a BorshSchemaContainer) -> Self {
        Self {
            container,
            generated: BTreeMap::new(),
        }
    }

    fn generate(&mut self) -> String {
        let mut output = "import { b } from 'zorsh';\n\n".to_string();

        // First pass - generate named structs
        for (name, def) in self.container.definitions() {
            if let Definition::Struct {
                fields: Fields::NamedFields(_),
            } = def
            {
                self.gen_schema(name, def);
            }
        }

        // Second pass - generate remaining types
        for (name, def) in self.container.definitions() {
            if !matches!(
                def,
                Definition::Struct {
                    fields: Fields::NamedFields(_)
                }
            ) {
                self.gen_schema(name, def);
            }
        }

        // Export named structs with types
        for (name, schema) in &self.generated {
            if let Some(Definition::Struct {
                fields: Fields::NamedFields(_),
            }) = self.container.get_definition(name)
            {
                output += &format!("export const {name}Schema = {schema};\n");
                output += &format!("export type {name} = b.infer<typeof {name}Schema>;\n\n");
            }
        }

        output
    }

    fn gen_schema(&mut self, name: &str, def: &Definition) {
        if self.generated.contains_key(name) {
            return;
        }

        let schema = match def {
            Definition::Primitive(size) => self.primitive_schema(*size, name),
            Definition::Sequence { elements, .. } => {
                format!("b.vec({})", self.resolve_type(elements))
            }
            Definition::Struct { fields } => self.struct_schema(fields),
            Definition::Enum { variants, .. } => self.enum_schema(variants),
            Definition::Tuple { elements } => {
                format!("b.tuple([{}])", self.resolve_types(elements))
            }
        };

        self.generated.insert(name.to_string(), schema);
    }

    fn resolve_type(&mut self, type_name: &str) -> String {
        if let Some(def) = self.container.get_definition(type_name) {
            self.gen_schema(type_name, def);
            if let Some(schema) = self.generated.get(type_name) {
                return schema.clone();
            }
        }
        self.primitive_schema(0, type_name)
    }

    fn resolve_types(&mut self, types: &[String]) -> String {
        types
            .iter()
            .map(|t| self.resolve_type(t))
            .collect::<Vec<_>>()
            .join(", ")
    }

    fn primitive_schema(&self, size: u8, name: &str) -> String {
        match (size, name) {
            (1, "bool") => "b.bool()".into(),
            (1, _) => "b.u8()".into(),
            (2, _) => "b.u16()".into(),
            (4, _) => "b.u32()".into(),
            (8, _) => "b.u64()".into(),
            (16, _) => "b.u128()".into(),
            (0, "()") => "b.unit()".into(),
            _ => panic!("Unsupported type: {}", name),
        }
    }

    fn struct_schema(&mut self, fields: &Fields) -> String {
        match fields {
            Fields::NamedFields(fields) => {
                let fields = fields
                    .iter()
                    .map(|(name, ty)| format!("{}: {}", name, self.resolve_type(ty)))
                    .collect::<Vec<_>>()
                    .join(", ");
                format!("b.struct({{ {} }})", fields)
            }
            Fields::UnnamedFields(fields) => format!("b.tuple([{}])", self.resolve_types(fields)),
            Fields::Empty => "b.unit()".into(),
        }
    }

    fn enum_schema(&mut self, variants: &[(i64, String, String)]) -> String {
        let variants = variants
            .iter()
            .map(|(_, name, ty)| {
                let schema = if ty == "()" {
                    "b.unit()".into()
                } else {
                    self.resolve_type(ty)
                };
                format!("{}: {}", name, schema)
            })
            .collect::<Vec<_>>()
            .join(", ");
        format!("b.enum({{ {} }})", variants)
    }
}

pub fn generate_zorsh_schema(schema_path: &str, output_path: &str) -> std::io::Result<()> {
    let mut schema_file = File::open(schema_path)?;
    let container = BorshSchemaContainer::deserialize_reader(&mut schema_file)?;

    let mut generator = SchemaGenerator::new(&container);
    let output = generator.generate();

    File::create(output_path)?.write_all(output.as_bytes())?;
    Ok(())
}

pub fn main() {
    let schema_path = "test_data/complex_schema.bin";
    let output_path = "generated_schema.ts";
    generate_zorsh_schema(schema_path, output_path).unwrap();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_zorsh_schema() {
        let schema_path = "test_data/complex_schema.bin";
        let output_path = "generated_schema.ts";
        generate_zorsh_schema(schema_path, output_path).unwrap();
    }
}
