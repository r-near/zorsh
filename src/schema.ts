import { BinaryReader, BinaryWriter } from "./binary-io"
import { type TypeRegistry, registry } from "./registry"
import type { VecType } from "./types"

export class Schema<T, Type extends string = string> {
  constructor(
    public readonly type: Type,
    public readonly options: unknown,
    private registry: TypeRegistry,
  ) {}

  serialize(value: T): Uint8Array {
    const writer = new BinaryWriter()
    const handler = this.registry.getHandler<T>(this.type)
    handler.write(writer, value, this.options)
    return writer.getBuffer()
  }

  deserialize(buffer: Uint8Array): T {
    const reader = new BinaryReader(buffer)
    const handler = this.registry.getHandler<T>(this.type)
    return handler.read(reader, this.options)
  }
}

// First declare the namespace type (for type-level stuff)
export namespace b {
  export type infer<T extends Schema<unknown>> = T extends Schema<infer U> ? U : never
}

// Builder API
export const b = {
  // Primitive types

  // Unsigned integers
  u8: () => new Schema<number, "u8">("u8", null, registry),
  u16: () => new Schema<number, "u16">("u16", null, registry),
  u32: () => new Schema<number, "u32">("u32", null, registry),
  u64: () => new Schema<bigint, "u64">("u64", null, registry),
  u128: () => new Schema<bigint, "u128">("u128", null, registry),

  // Signed integers
  i8: () => new Schema<number, "i8">("i8", null, registry),
  i16: () => new Schema<number, "i16">("i16", null, registry),
  i32: () => new Schema<number, "i32">("i32", null, registry),
  i64: () => new Schema<bigint, "i64">("i64", null, registry),
  i128: () => new Schema<bigint, "i128">("i128", null, registry),

  // Floating point
  f32: () => new Schema<number, "f32">("f32", null, registry),
  f64: () => new Schema<number, "f64">("f64", null, registry),

  // Other primitives
  string: () => new Schema<string, "string">("string", null, registry),
  unit: () => new Schema<Record<string, never>, "unit">("unit", null, registry),

  // Complex container types
  vec: <T extends Schema<unknown>>(elementSchema: T): Schema<VecType<T>> => {
    return new Schema(
      "vec",
      {
        elementType: elementSchema.type,
        elementOptions: elementSchema.options,
      },
      registry,
    )
  },

  hashSet: <T>(elementSchema: Schema<T>): Schema<Set<T>> => {
    return new Schema(
      "set",
      {
        elementType: elementSchema.type,
        elementOptions: elementSchema.options,
      },
      registry,
    )
  },

  hashMap: <K, V>(keySchema: Schema<K>, valueSchema: Schema<V>): Schema<Map<K, V>> => {
    return new Schema(
      "map",
      {
        keyType: keySchema.type,
        keyOptions: keySchema.options,
        valueType: valueSchema.type,
        valueOptions: valueSchema.options,
      },
      registry,
    )
  },

  option: <T>(valueSchema: Schema<T>): Schema<T | null> => {
    return new Schema(
      "option",
      {
        valueType: valueSchema.type,
        valueOptions: valueSchema.options,
      },
      registry,
    )
  },

  array: <T>(elementSchema: Schema<T>, length: number): Schema<T[]> => {
    return new Schema(
      "array",
      {
        elementType: elementSchema.type,
        elementOptions: elementSchema.options,
        length,
      },
      registry,
    )
  },

  // Enum type
  enum: <T extends Record<string, Schema<unknown>>>(
    variants: T,
  ): Schema<
    {
      [K in keyof T]: { [KK in K]: T[K] extends Schema<infer U> ? U : never }
    }[keyof T]
  > => {
    // Convert variants to array format with indices
    const variantArray = Object.entries(variants).map(([name, schema], index) => ({
      index,
      name,
      type: schema.type,
      options: schema.options,
    }))

    return new Schema("enum", { variants: variantArray }, registry)
  },

  // Complex types
  struct: <T extends Record<string, Schema<unknown>>>(
    fields: T,
  ): Schema<{ [K in keyof T]: T[K] extends Schema<infer U> ? U : never }> => {
    const fieldDefs = Object.fromEntries(
      Object.entries(fields).map(([key, schema]) => [
        key,
        { type: schema.type, options: schema.options },
      ]),
    )
    return new Schema("struct", fieldDefs, registry)
  },
}
