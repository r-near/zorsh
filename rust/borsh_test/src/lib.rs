// src/lib.rs
use borsh::{BorshDeserialize, BorshSerialize};
use std::collections::{HashMap, HashSet};

mod float_test;

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub enum GameEvent {
    PlayerJoined {
        player_id: String,
    },
    PlayerLeft {
        player_id: String,
    },
    ChatMessage {
        sender: String,
        message: String,
    },
    ItemUsed {
        item_id: String,
        target: Option<String>,
    }, // Option for single-target or area-effect
    BattleResult {
        winner: String,
        loser: String,
        rewards: Vec<Reward>, // Nested struct
    },
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct Reward {
    item_id: String,
    quantity: u32,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct Player {
    pub name: String,
    pub level: u8,
    pub stats: Stats,                            // Nested struct
    pub inventory: Vec<Item>,                    // Vec of nested structs
    pub equipped_items: HashMap<String, String>, // Slot -> Item ID
    pub quest_log: HashSet<String>,              // Set of quest IDs
    pub last_login: Option<u64>,                 // Optional timestamp
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct Stats {
    pub health: u32,
    pub mana: u32,
    pub attack: u32,
    pub defense: u32,
    pub magic_attack: u32,
    pub magic_defense: u32,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct Item {
    pub id: String,
    pub name: String,
    pub weight: f32,
    pub effects: Vec<Effect>, // Vec of nested enum
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub enum Effect {
    Damage(u32),
    Heal(u32),
    Buff(StatBuff),   // Nested struct
    Debuff(StatBuff), // Nested struct
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct StatBuff {
    pub stat: String, // Could be an enum, but string for simplicity
    pub amount: i32,  // Can be positive or negative
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct GameState {
    pub players: HashMap<String, Player>, // Map of player ID to Player struct
    pub current_round: u32,
    pub events: Vec<GameEvent>, // Vec of enum
    pub game_version: String,
}

#[cfg(test)]
mod tests {
    use super::*;
    use borsh::to_vec;
    use borsh::BorshSerialize;
    use std::fs::File;
    use std::io::Write;

    fn write_test_data<T: BorshSerialize>(
        data: &T,
        base_path: &str,
        name: &str,
    ) -> Result<(), std::io::Error> {
        let bin_path = format!("{}/{}.bin", base_path, name);
        let serialized = to_vec(data)?;
        let mut bin_file = File::create(bin_path)?;
        bin_file.write_all(&serialized)?;
        Ok(())
    }

    #[test]
    #[ignore]
    fn generate_test_data() {
        let test_data_dir = "test_data";
        std::fs::create_dir_all(test_data_dir).expect("Failed to create test_data dir");

        // --- Create some sample data ---

        let mut players = HashMap::new();

        let player1 = Player {
            name: "Alice".to_string(),
            level: 10,
            stats: Stats {
                health: 100,
                mana: 50,
                attack: 20,
                defense: 15,
                magic_attack: 5,
                magic_defense: 10,
            },
            inventory: vec![
                Item {
                    id: "sword_001".to_string(),
                    name: "Iron Sword".to_string(),
                    weight: 2.5,
                    effects: vec![Effect::Damage(15)],
                },
                Item {
                    id: "potion_001".to_string(),
                    name: "Health Potion".to_string(),
                    weight: 0.125,
                    effects: vec![Effect::Heal(50)],
                },
            ],
            equipped_items: HashMap::from([("hand".to_string(), "sword_001".to_string())]),
            quest_log: HashSet::from(["quest_start".to_string(), "collect_herbs".to_string()]),
            last_login: Some(1678886400), // Example timestamp
        };

        let player2 = Player {
            name: "Bob".to_string(),
            level: 5,
            stats: Stats {
                health: 80,
                mana: 75,
                attack: 10,
                defense: 5,
                magic_attack: 30,
                magic_defense: 20,
            },
            inventory: vec![Item {
                id: "staff_001".to_string(),
                name: "Apprentice Staff".to_string(),
                weight: 1.25,
                effects: vec![
                    Effect::Damage(5),
                    Effect::Buff(StatBuff {
                        stat: "magic_attack".to_string(),
                        amount: 10,
                    }),
                ],
            }],
            equipped_items: HashMap::new(),
            quest_log: HashSet::new(),
            last_login: None,
        };

        players.insert("alice_id".to_string(), player1);
        players.insert("bob_id".to_string(), player2);

        let events = vec![
            GameEvent::PlayerJoined {
                player_id: "alice_id".to_string(),
            },
            GameEvent::ChatMessage {
                sender: "alice_id".to_string(),
                message: "Hello, world!".to_string(),
            },
            GameEvent::BattleResult {
                winner: "alice_id".to_string(),
                loser: "bob_id".to_string(),
                rewards: vec![Reward {
                    item_id: "gold_001".to_string(),
                    quantity: 100,
                }],
            },
        ];

        let game_state = GameState {
            players,
            current_round: 3,
            events,
            game_version: "1.2.3".to_string(),
        };

        write_test_data(&game_state, test_data_dir, "complex_game_state")
            .expect("Failed to write game state data");
    }
}
