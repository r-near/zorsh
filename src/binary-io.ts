// Core binary reading/writing utilities

export class BinaryWriter {
	private buffer: Uint8Array;
	private view: DataView;
	private offset = 0;

	constructor(initialSize = 1024) {
		this.buffer = new Uint8Array(initialSize);
		this.view = new DataView(this.buffer.buffer);
	}

	private ensureCapacity(additionalBytes: number): void {
		const requiredSize = this.offset + additionalBytes;
		if (requiredSize <= this.buffer.length) {
			return;
		}

		// Double the buffer size until it's large enough
		let newSize = this.buffer.length;
		while (newSize < requiredSize) {
			newSize *= 2;
		}

		const newBuffer = new Uint8Array(newSize);
		newBuffer.set(this.buffer);
		this.buffer = newBuffer;
		this.view = new DataView(this.buffer.buffer);
	}

	// Unsigned integers
	writeUint8(value: number): void {
		this.ensureCapacity(1);
		this.view.setUint8(this.offset, value);
		this.offset += 1;
	}

	writeUint16(value: number): void {
		this.ensureCapacity(2);
		this.view.setUint16(this.offset, value, true);
		this.offset += 2;
	}

	writeUint32(value: number): void {
		this.ensureCapacity(4);
		this.view.setUint32(this.offset, value, true);
		this.offset += 4;
	}

	writeUint64(value: bigint): void {
		this.ensureCapacity(8);
		this.view.setBigUint64(this.offset, value, true);
		this.offset += 8;
	}

	writeUint128(value: bigint): void {
		this.ensureCapacity(16);
		const low = value & BigInt("0xFFFFFFFFFFFFFFFF");
		const high = value >> BigInt(64);
		this.writeUint64(low);
		this.writeUint64(high);
	}

	// Signed integers
	writeInt8(value: number): void {
		this.ensureCapacity(1);
		this.view.setInt8(this.offset, value);
		this.offset += 1;
	}

	writeInt16(value: number): void {
		this.ensureCapacity(2);
		this.view.setInt16(this.offset, value, true);
		this.offset += 2;
	}

	writeInt32(value: number): void {
		this.ensureCapacity(4);
		this.view.setInt32(this.offset, value, true);
		this.offset += 4;
	}

	writeInt64(value: bigint): void {
		this.ensureCapacity(8);
		this.view.setBigInt64(this.offset, value, true);
		this.offset += 8;
	}

	writeInt128(value: bigint): void {
		this.ensureCapacity(16);
		const low = value & BigInt("0xFFFFFFFFFFFFFFFF");
		const high = value >> BigInt(64);
		this.writeInt64(low);
		this.writeInt64(high);
	}

	// Floating point
	writeFloat32(value: number): void {
		this.ensureCapacity(4);
		this.view.setFloat32(this.offset, value, true);
		this.offset += 4;
	}

	writeFloat64(value: number): void {
		this.ensureCapacity(8);
		this.view.setFloat64(this.offset, value, true);
		this.offset += 8;
	}

	writeString(value: string): void {
		const bytes = new TextEncoder().encode(value);
		this.writeUint32(bytes.length);
		this.ensureCapacity(bytes.length);
		this.buffer.set(bytes, this.offset);
		this.offset += bytes.length;
	}

	getBuffer(): Uint8Array {
		return this.buffer.slice(0, this.offset);
	}
}

export class BinaryReader {
	private view: DataView;
	private offset = 0;

	constructor(private buffer: Uint8Array) {
		this.view = new DataView(
			buffer.buffer,
			buffer.byteOffset,
			buffer.byteLength,
		);
	}

	// Unsigned integers
	readUint8(): number {
		const value = this.view.getUint8(this.offset);
		this.offset += 1;
		return value;
	}

	readUint16(): number {
		const value = this.view.getUint16(this.offset, true);
		this.offset += 2;
		return value;
	}

	readUint32(): number {
		const value = this.view.getUint32(this.offset, true);
		this.offset += 4;
		return value;
	}

	readUint64(): bigint {
		const value = this.view.getBigUint64(this.offset, true);
		this.offset += 8;
		return value;
	}

	readUint128(): bigint {
		const low = this.readUint64();
		const high = this.readUint64();
		return (high << BigInt(64)) | low;
	}

	// Signed integers
	readInt8(): number {
		const value = this.view.getInt8(this.offset);
		this.offset += 1;
		return value;
	}

	readInt16(): number {
		const value = this.view.getInt16(this.offset, true);
		this.offset += 2;
		return value;
	}

	readInt32(): number {
		const value = this.view.getInt32(this.offset, true);
		this.offset += 4;
		return value;
	}

	readInt64(): bigint {
		const value = this.view.getBigInt64(this.offset, true);
		this.offset += 8;
		return value;
	}

	readInt128(): bigint {
		const low = this.readUint64(); // Read as unsigned to preserve bit pattern
		const high = this.readInt64(); // Read high bits as signed
		return (high << BigInt(64)) | low;
	}

	// Floating point
	readFloat32(): number {
		const value = this.view.getFloat32(this.offset, true);
		this.offset += 4;
		return value;
	}

	readFloat64(): number {
		const value = this.view.getFloat64(this.offset, true);
		this.offset += 8;
		return value;
	}

	readString(): string {
		const length = this.readUint32();
		const bytes = this.buffer.slice(this.offset, this.offset + length);
		this.offset += length;
		return new TextDecoder().decode(bytes);
	}

	getOffset(): number {
		return this.offset;
	}
}
