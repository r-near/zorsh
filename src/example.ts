import { b } from "./schema";

// Define a schema
const UserSchema = b.struct({
	name: b.string(),
	age: b.u8(),
	scores: b.vec(b.u16()),
	metadata: b.option(
		b.struct({
			joinDate: b.u64(),
			rank: b.string(),
		}),
	),
});

// Get the type from the schema using b.infer
type User = b.infer<typeof UserSchema>;

// TypeScript infers this as:
// type User = {
//   name: string
//   age: number
//   scores: number[]
//   metadata: {
//     joinDate: bigint
//     rank: string
//   } | null
// }

// Use the inferred type
const user: User = {
	name: "Alice",
	age: 25,
	scores: [100, 95, 98],
	metadata: {
		joinDate: BigInt(Date.now()),
		rank: "Gold",
	},
};

const bytes = UserSchema.serialize(user);
const decoded = UserSchema.deserialize(bytes);
console.log("Decoded:", decoded);
