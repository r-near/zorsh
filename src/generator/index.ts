export * from "./types"
export * from "./generator"
export * from "./analyzer"
export * from "./resolver"
export * from "./parser"
