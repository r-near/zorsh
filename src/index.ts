export { b, Schema } from "./schema"
export type { EnumLike, EnumValueType } from "./types"
