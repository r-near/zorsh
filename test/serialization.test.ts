import { describe, expect, test } from "vitest"
import { type Schema, b } from "../src/schema"

interface TestType<T> {
  schema: Schema<T>
  value: T
}

describe("primitive types", () => {
  test.each([
    { schema: b.u8(), value: 255 },
    { schema: b.u16(), value: 65535 },
    { schema: b.u32(), value: 4294967295 },
    { schema: b.u64(), value: BigInt("18446744073709551615") },
    {
      schema: b.u128(),
      value: BigInt("340282366920938463463374607431768211455"),
    },
    { schema: b.i8(), value: -128 },
    { schema: b.i16(), value: -32768 },
    { schema: b.i32(), value: -2147483648 },
    { schema: b.i64(), value: BigInt("-9223372036854775808") },
    {
      schema: b.i128(),
      value: BigInt("-170141183460469231731687303715884105728"),
    },
    { schema: b.f32(), value: 0.015625 },
    { schema: b.f64(), value: Math.PI },
  ])("numeric type $schema.type", ({ schema, value }: TestType<number | bigint>) => {
    const buffer = schema.serialize(value)
    expect(buffer).toMatchSnapshot()
    expect(schema.deserialize(buffer)).toEqual(value)
  })

  test.each([
    { schema: b.bool(), value: true },
    { schema: b.bool(), value: false },
  ])("bool", ({ schema, value }: TestType<boolean>) => {
    const buffer = schema.serialize(value)
    expect(buffer).toMatchSnapshot()
    expect(schema.deserialize(buffer)).toEqual(value)
  })

  test("string", () => {
    const schema = b.string()
    const value = "Hello, world! ðŸ‘‹"
    const buffer = schema.serialize(value)
    expect(buffer).toMatchSnapshot()
    expect(schema.deserialize(buffer)).toBe(value)
  })

  test("unit", () => {
    const schema = b.unit()
    const value = {}
    const buffer = schema.serialize(value)
    expect(buffer).toMatchSnapshot()
    expect(schema.deserialize(buffer)).toEqual(value)
  })
})

describe("complex types", () => {
  test("fixed array", () => {
    const schema = b.array(b.u16(), 3)
    const value = [1, 2, 3]
    const buffer = schema.serialize(value)
    expect(buffer).toMatchSnapshot()
    expect(schema.deserialize(buffer)).toEqual(value)

    // Should throw on wrong length
    expect(() => schema.serialize([1, 2])).toThrow()
  })

  test("dynamic vec", () => {
    const schema = b.vec(b.u8())
    const value = new Uint8Array([1, 2, 3, 4, 5])
    const buffer = schema.serialize(value)
    expect(buffer).toMatchSnapshot()
    expect(schema.deserialize(buffer)).toEqual(value)
  })

  test("option", () => {
    const schema = b.option(b.string())

    const someValue = "hello"
    const someBuffer = schema.serialize(someValue)
    expect(someBuffer).toMatchSnapshot("some value")
    expect(schema.deserialize(someBuffer)).toBe(someValue)

    const noneValue = null
    const noneBuffer = schema.serialize(noneValue)
    expect(noneBuffer).toMatchSnapshot("none value")
    expect(schema.deserialize(noneBuffer)).toBe(noneValue)
  })

  test("hashset", () => {
    const schema = b.hashSet(b.string())
    const value = new Set(["a", "b", "c"])
    const buffer = schema.serialize(value)
    expect(buffer).toMatchSnapshot()
    expect(schema.deserialize(buffer)).toEqual(value)
  })

  test("hashmap", () => {
    const schema = b.hashMap(b.string(), b.u32())
    const value = new Map([
      ["alice", 100],
      ["bob", 200],
    ])
    const buffer = schema.serialize(value)
    expect(buffer).toMatchSnapshot()
    expect(schema.deserialize(buffer)).toEqual(value)
  })

  test("struct", () => {
    const schema = b.struct({
      name: b.string(),
      age: b.u8(),
      scores: b.vec(b.u16()),
    })

    const value = {
      name: "alice",
      age: 25,
      scores: new Uint16Array([90, 95, 100]),
    }

    const buffer = schema.serialize(value)
    expect(buffer).toMatchSnapshot()
    expect(schema.deserialize(buffer)).toEqual(value)
  })

  test("enum", () => {
    // Simple enum
    const StatusSchema = b.enum({
      Pending: b.unit(),
      Fulfilled: b.unit(),
      Rejected: b.unit(),
    })

    const statusValue = { Pending: {} }
    const statusBuffer = StatusSchema.serialize(statusValue)
    expect(statusBuffer).toMatchSnapshot("simple enum")
    expect(StatusSchema.deserialize(statusBuffer)).toEqual(statusValue)

    // Complex enum
    const ResultSchema = b.enum({
      Ok: b.string(),
      Err: b.struct({
        code: b.u16(),
        message: b.string(),
      }),
    })

    const okValue = { Ok: "success" }
    const okBuffer = ResultSchema.serialize(okValue)
    expect(okBuffer).toMatchSnapshot("enum with data - ok")
    expect(ResultSchema.deserialize(okBuffer)).toEqual(okValue)

    const errValue = { Err: { code: 404, message: "Not found" } }
    const errBuffer = ResultSchema.serialize(errValue)
    expect(errBuffer).toMatchSnapshot("enum with data - err")
    expect(ResultSchema.deserialize(errBuffer)).toEqual(errValue)
  })
})

describe("complex nested structures", () => {
  test("game state example", () => {
    const PlayerSchema = b.struct({
      name: b.string(),
      health: b.u8(),
      inventory: b.hashMap(b.string(), b.u16()),
      achievements: b.hashSet(b.string()),
      guild: b.option(b.string()),
      recentScores: b.vec(b.u16()),
    })

    const GameStateSchema = b.struct({
      players: b.hashMap(b.string(), PlayerSchema),
      currentRound: b.u32(),
      nextRoundStart: b.option(b.u32()),
    })

    const value = {
      players: new Map([
        [
          "player1",
          {
            name: "Alice",
            health: 100,
            inventory: new Map([
              ["sword", 1],
              ["potion", 5],
            ]),
            achievements: new Set(["first_kill", "speedrun"]),
            guild: "Warriors",
            recentScores: new Uint16Array([100, 95, 98]),
          },
        ],
      ]),
      currentRound: 5,
      nextRoundStart: 1234567890,
    }

    const buffer = GameStateSchema.serialize(value)
    expect(buffer).toMatchSnapshot()
    expect(GameStateSchema.deserialize(buffer)).toEqual(value)
  })
})

describe("type inference", () => {
  test("type inference matches runtime values", () => {
    const UserSchema = b.struct({
      name: b.string(),
      age: b.u8(),
      scores: b.vec(b.u16()),
      metadata: b.option(
        b.struct({
          joinDate: b.u64(),
          rank: b.string(),
        }),
      ),
    })

    type User = b.infer<typeof UserSchema>

    // This is a type-level test - if it compiles, it passes
    const user: User = {
      name: "Alice",
      age: 25,
      scores: new Uint16Array([100, 95, 98]),
      metadata: {
        joinDate: BigInt(Date.now()),
        rank: "Gold",
      },
    }

    // These should fail type checking:
    // @ts-expect-error - wrong age type
    // biome-ignore lint/correctness/noUnusedVariables: This is a test
    const user1: User = { ...user, age: "25" }

    // @ts-expect-error - missing field
    // biome-ignore lint/correctness/noUnusedVariables: This is a test
    const user2: User = { name: "Bob", age: 30 }

    // @ts-expect-error - wrong metadata structure
    // biome-ignore lint/correctness/noUnusedVariables: This is a test
    const user3: User = { ...user, metadata: { rank: "Silver" } }

    // Runtime test
    const buffer = UserSchema.serialize(user)
    const decoded = UserSchema.deserialize(buffer)
    expect(decoded).toEqual(user)
  })
})
