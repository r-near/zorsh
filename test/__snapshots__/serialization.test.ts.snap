// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`complex nested structures > game state example 1`] = `
Uint8Array [
  1,
  0,
  0,
  0,
  7,
  0,
  0,
  0,
  112,
  108,
  97,
  121,
  101,
  114,
  49,
  5,
  0,
  0,
  0,
  65,
  108,
  105,
  99,
  101,
  100,
  2,
  0,
  0,
  0,
  6,
  0,
  0,
  0,
  112,
  111,
  116,
  105,
  111,
  110,
  5,
  0,
  5,
  0,
  0,
  0,
  115,
  119,
  111,
  114,
  100,
  1,
  0,
  2,
  0,
  0,
  0,
  10,
  0,
  0,
  0,
  102,
  105,
  114,
  115,
  116,
  95,
  107,
  105,
  108,
  108,
  8,
  0,
  0,
  0,
  115,
  112,
  101,
  101,
  100,
  114,
  117,
  110,
  1,
  8,
  0,
  0,
  0,
  87,
  97,
  114,
  114,
  105,
  111,
  114,
  115,
  3,
  0,
  0,
  0,
  100,
  0,
  95,
  0,
  98,
  0,
  5,
  0,
  0,
  0,
  1,
  210,
  2,
  150,
  73,
]
`;

exports[`complex types > dynamic vec 1`] = `
Uint8Array [
  5,
  0,
  0,
  0,
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`complex types > enum > enum with data - err 1`] = `
Uint8Array [
  1,
  148,
  1,
  9,
  0,
  0,
  0,
  78,
  111,
  116,
  32,
  102,
  111,
  117,
  110,
  100,
]
`;

exports[`complex types > enum > enum with data - ok 1`] = `
Uint8Array [
  0,
  7,
  0,
  0,
  0,
  115,
  117,
  99,
  99,
  101,
  115,
  115,
]
`;

exports[`complex types > enum > simple enum 1`] = `
Uint8Array [
  0,
]
`;

exports[`complex types > fixed array 1`] = `
Uint8Array [
  1,
  0,
  2,
  0,
  3,
  0,
]
`;

exports[`complex types > hashmap 1`] = `
Uint8Array [
  2,
  0,
  0,
  0,
  5,
  0,
  0,
  0,
  97,
  108,
  105,
  99,
  101,
  100,
  0,
  0,
  0,
  3,
  0,
  0,
  0,
  98,
  111,
  98,
  200,
  0,
  0,
  0,
]
`;

exports[`complex types > hashset 1`] = `
Uint8Array [
  3,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  97,
  1,
  0,
  0,
  0,
  98,
  1,
  0,
  0,
  0,
  99,
]
`;

exports[`complex types > option > none value 1`] = `
Uint8Array [
  0,
]
`;

exports[`complex types > option > some value 1`] = `
Uint8Array [
  1,
  5,
  0,
  0,
  0,
  104,
  101,
  108,
  108,
  111,
]
`;

exports[`complex types > struct 1`] = `
Uint8Array [
  5,
  0,
  0,
  0,
  97,
  108,
  105,
  99,
  101,
  25,
  3,
  0,
  0,
  0,
  90,
  0,
  95,
  0,
  100,
  0,
]
`;

exports[`primitive types > numeric type 'f32' 1`] = `
Uint8Array [
  0,
  0,
  128,
  60,
]
`;

exports[`primitive types > numeric type 'f64' 1`] = `
Uint8Array [
  24,
  45,
  68,
  84,
  251,
  33,
  9,
  64,
]
`;

exports[`primitive types > numeric type 'i8' 1`] = `
Uint8Array [
  128,
]
`;

exports[`primitive types > numeric type 'i16' 1`] = `
Uint8Array [
  0,
  128,
]
`;

exports[`primitive types > numeric type 'i32' 1`] = `
Uint8Array [
  0,
  0,
  0,
  128,
]
`;

exports[`primitive types > numeric type 'i64' 1`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  128,
]
`;

exports[`primitive types > numeric type 'i128' 1`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  128,
]
`;

exports[`primitive types > numeric type 'u8' 1`] = `
Uint8Array [
  255,
]
`;

exports[`primitive types > numeric type 'u16' 1`] = `
Uint8Array [
  255,
  255,
]
`;

exports[`primitive types > numeric type 'u32' 1`] = `
Uint8Array [
  255,
  255,
  255,
  255,
]
`;

exports[`primitive types > numeric type 'u64' 1`] = `
Uint8Array [
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
]
`;

exports[`primitive types > numeric type 'u128' 1`] = `
Uint8Array [
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
]
`;

exports[`primitive types > string 1`] = `
Uint8Array [
  18,
  0,
  0,
  0,
  72,
  101,
  108,
  108,
  111,
  44,
  32,
  119,
  111,
  114,
  108,
  100,
  33,
  32,
  240,
  159,
  145,
  139,
]
`;

exports[`primitive types > unit 1`] = `Uint8Array []`;
